docs/architecture/adr-048-consensus-fees.md

```diff
@@ -0,0 +1,194 @@
+For tier 2 and tier 3 transactions, the gas price is adjusted according to previous block load, the logic could be similar to EIP-1559:
+
+```python
+def adjust_gas_price(gas_price, parent_gas_used, tier):
+  if parent_gas_used == tier.parent_gas_target:
+    return gas_price
+  elif parent_gas_used > tier.parent_gas_target:
+    gas_used_delta = parent_gas_used - tier.parent_gas_target
+    gas_price_delta = max(gas_price * gas_used_delta // tier.parent_gas_target // tier.change_speed, 1)
+    return gas_price + gas_price_delta
+  else:
+    gas_used_delta = parent_gas_target - parent_gas_used
+    gas_price_delta = gas_price * gas_used_delta // parent_gas_target // tier.change_speed
+    return gas_price - gas_price_delta
+```
+
+### Block Segment Reservation
+
+Ideally we should reserve block segments for each tier, so the lower tiered transactions won't be completely squeezed out by higher tier transactions, which will force user to use higher tier, and the system degraded to a single tier.
+
+We need help from tendermint to implement this.
+
+## Implementation
+
+We can make each tier's gas price strategy fully configurable in protocol parameters, while providing a sensible default one.
+
+Pseudocode in python-like syntax:
+
+```python
+interface TieredTx:
+  def tier(self) -> int:
+    pass
+
+def tx_tier(tx):
+    if isinstance(tx, TieredTx):
+      return tx.tier()
+    else:
+      # default tier for custom transactions
+      return 0
+
+class TierParams:
+  'gas price strategy parameters of one tier'
+  priority: int           # priority in tendermint mempool
+  initial_gas_price: Coin
+  parent_gas_target: int
+  change_speed: Decimal   # 0 means don't adjust for block load.
+
+class Params:
+    'protocol parameters'
+    tiers: List[TierParams]
+
+class State:
+    'consensus state'
+    # total gas used in last block, None when it's the first block
+    parent_gas_used: Optional[int]
+    # gas prices of last block for all tiers
+    gas_prices: List[Coin]
+
+def begin_block():
+    'Adjust gas prices'
+    for i, tier in enumerate(Params.tiers):
+        if State.parent_gas_used is None:
+            # initialized gas price for the first block
+	          State.gas_prices[i] = tier.initial_gas_price
+        else:
+            # adjust gas price according to gas used in previous block
+            State.gas_prices[i] = adjust_gas_price(State.gas_prices[i], State.parent_gas_used, tier)
+
+def mempoolFeeTxHandler_checkTx(ctx, tx):
+    # the minimal-gas-price configured by validator, zero in deliver_tx context
+    validator_price = ctx.MinGasPrice()
+    consensus_price = State.gas_prices[tx_tier(tx)]
+    min_price = max(validator_price, consensus_price)
+
+    if tx.gas_price() < min_price:
+        return 'insufficient fees'
+    return next_CheckTx(ctx, tx)
+
+def txPriorityHandler_checkTx(ctx, tx):
+    res, err := next_CheckTx(ctx, tx)
+    # pass priority to tendermint
+    res.Priority = Params.tiers[tx_tier(tx)].priority
```

Comment:
`tx_tire` returns number, so we should remove  `.priority`.
```suggestion
    res.Priority = Params.tiers[tx_tier(tx)]
```

Reply:
hmm the original one is good, `tx_tier(tx)` returns the `gas_tier`, then it's used to tier parameter list `Params.tiers` to get the tier parameters, then use the configured `.priority` field of it as the tx priority.